### PR TITLE
Disable graph analysis at query time for shingle and cjk filters producing tokens of different size

### LINKED ISSUE #23918: Shingle filters that produce shingles of different size can create gigantic queries
Shingle filters creates a graph token stream that the query parser is now able to consume.
Though when shingles of different size are produced the number of paths in the graph can explode.
This is also the case when `output_unigram` is set to true.
In 5.3 all paths are generated before building the query so a node can OOM easily on a single big input query. In 5.4 and beyond we detect the explosion earlier but we fail the entire request.
Instead we should be able to detect the problematic token filters and disable the graph analysis for these fields.

### PR DESCRIPTION
**Pull request by jimczi**
Shingle filters that produce shingles of different size and CJK filters that produce bigram AND unigram are problematic when
 we analyze the graph they produce. The position for each shingle size are not aligned so each position has at least two side paths.
 So in order to avoid paths explosion this change disables the graph analysis at query time for field analyzers that contain these filters
 with a problematic configuration.

Closes #23918

### CODE REVIEW

File: core/src/test/java/org/elasticsearch/index/query/DisableGraphQueryTests.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        assertThat(expectedQuery, equalTo(query));
+    }
+
+    public void testQueryString() throws IOException {
+        QueryStringQueryBuilder builder = new QueryStringQueryBuilder("foo bar baz");
+        builder.field("text_shingle");
+        builder.splitOnWhitespace(false);
+        Query query = builder.doToQuery(shardContext);
+        assertThat(expectedQuery, equalTo(query));
+    }
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/index/query/DisableGraphQueryTests.java]: should we also check that the correct query is built if the analyzer is sane and does not output unigrams?

**Review by jpountz — COMMENTED** [review]
It looks good. I'm wondering we should also test that we keep processing the query like before if the shingles are sane?

**Comment by jimczi** [comment]
@jpountz I pushed some changes to address your comment.
I also moved the logic to disable graph analysis at a higher level in order to handle phrase queries too.

**Review by jpountz — APPROVED** [review]
LGTM

**Pull request merged**